Write Read, Replace, Patch BatchV1NamespacedJobStatus test - +3 endpoint coverage

riaankleinhans opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceBatchV1NamespacedJobStatus
- readBatchV1NamespacedJobStatus
- patchBatchV1NamespacedJobStatus 

**Which issue(s) this PR fixes:**
Fixes #108113

**Testgrid Link:**
[testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20apply%20changes%20to%20a%20job%20status&graph-metrics=test-duration-minutes)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

Riaankl commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

Unrelater flakes
```
[sig-node] Probing container
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/node/framework.go:23
  should have monotonically increasing restart count [NodeConformance] [Conformance] [It]
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633
  
  [sig-network] Proxy
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/common/framework.go:23
  version v1
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/proxy.go:74
    A set of valid responses are returned for both pod and service ProxyWithPath [Conformance] [It]
    /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633
    
```

dims commented:
/assign @janetkuo @soltysh

Riaankl commented:
/assign @alculquicondor

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +366,64 @@ var _ = SIGDescribe("Job", func() {
+
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJob("notTerminate", "suspend-false-to-true", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit)
+		job.Spec.Suspend = pointer.BoolPtr(false)
```
no need for this line (it's false by default)

heyste replied:
Removed.

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +366,64 @@ var _ = SIGDescribe("Job", func() {
+				return err
+			}
+			statusToUpdate.Status.StartTime = &now2
+			updatedStatus, err = jClient.UpdateStatus(context.TODO(), statusToUpdate, metav1.UpdateOptions{})
```
The job controller already updates the status. Is that not detected by the APISnoop?

heyste replied:
APISnoop uses the e2e test name, in this case "should verify changes to a job status" as part of the _useragent_ to track which endpoints have been hit. Indirect calls by the controller can't be associated with this test to validate that an endpoint was hit.

This pattern has been used to test status endpoints for other resources. e.g. [Write Service Status Life Cycle test - +4 endpoint coverage](https://github.com/kubernetes/kubernetes/pull/98018#)

alculquicondor replied:
got it, thanks

heyste commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
unrelated flake `[sig-instrumentation] MetricsGrabber should grab all metrics from a Kubelet.`

```
Status: "Failure",
Message: "the server is currently unable to handle the request (get nodes e2e-2aeef39800-a7d53-minion-group-4hkw:10250)",
Reason: "ServiceUnavailable",
```

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +366,63 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
+
+	ginkgo.It("should verify changes to a job status", func() {
```
```suggestion
	ginkgo.It("should apply changes to a job status", func() {
```

heyste replied:
Fixed

alculquicondor reviewed (commented):
LGTM after the minor change

heyste commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel
unrelated flake
```
s: "timed out waiting for the condition",
```

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +366,63 @@ var _ = SIGDescribe("Job", func() {
+
+		ginkgo.By("get /status")
+		jResource := schema.GroupVersionResource{Group: "batch", Version: "v1", Resource: "jobs"}
+		gottenStatus, err := f.DynamicClient.Resource(jResource).Namespace(ns).Get(context.TODO(), job.Name, metav1.GetOptions{}, "status")
```
oh, I just noticed this. Any particular reason to use a DynamicClient instead of a regular client?

Also, it doesn't look like you are querying the status subresource.

Lastly, `gotStatus` is a more common variable name used in tests.

heyste replied:
As naming is hard my preference is to use the current style that the sig has been using. When testing `get /status`, other than calling the `job` resource over `cronjob` this test follows the same pattern as `cronjob.go`

https://github.com/kubernetes/kubernetes/blob/80746c762227a19fd4588e39b93eb9049dbff8ce/test/e2e/apps/cronjob.go#L423-L429

Querying apisnoop

```
select distinct  endpoint, right(useragent,41) AS useragent
from testing.audit_event
where endpoint ilike '%JobStatus%'
and useragent like 'e2e%should%'
order by endpoint;
```
shows the following endpoints are hit
```
             endpoint              |                 useragent
-----------------------------------+-------------------------------------------
 patchBatchV1NamespacedJobStatus   |  Job should apply changes to a job status
 readBatchV1NamespacedJobStatus    |  Job should apply changes to a job status
 replaceBatchV1NamespacedJobStatus |  Job should apply changes to a job status

```

alculquicondor replied:
Oh, I missed the `"status")` at the end of the line.

heyste committed:
Create e2e job status test

e2e test validates the following 3 extra endpoints
- patchBatchV1NamespacedJobStatus
- readBatchV1NamespacedJobStatus
- replaceBatchV1NamespacedJobStatus

heyste commented:
/test pull-kubernetes-node-e2e-containerd
unrelated flake
```
[sig-node] Summary API [NodeConformance]
test/e2e_node/framework.go:23
  when querying /stats/summary
  test/e2e_node/summary_test.go:42
    should report resource usage through the stats api [It]
    test/e2e_node/summary_test.go:53

    Timed out after 180.000s.
```

heyste commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
unrelated flake
```
[sig-storage] CSI Volumes [Driver: csi-hostpath]  [Testpattern: Generic Ephemeral-volume (default fs)] volumeLimits should  verify that all csinodes have volume limits 
{ Failure test/e2e/storage/testsuites/volumelimits.go:247 Mar 21 22:11:25.356: Expected volume limits to be set, error: could not get CSINode limit for driver csi-hostpath: timed out waiting for the condition vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113}
```

dims commented:
/approve
/lgtm
